state/protocol/badger/state.go

```diff
@@ -14,6 +14,29 @@ import (
 	"github.com/onflow/flow-go/storage/badger/operation"
 )
 
+type MutatorFactory interface {
+	Create(state *State) protocol.Mutator
+}
+
+type mutatorFactory struct {
+	results storage.ExecutionResults
+}
+
+func (m *mutatorFactory) Create(state *State) protocol.Mutator {
+	r := &Mutator{
+		state:     state,
+		validator: NewReceiptValidator(state, state.index, m.results),
```

Comment:
Do we need to instantiate a new `ReceiptValidator` each time? I believe we should be able to instantiate one instance when constructing the `State` and use that. The `protocol.State` and all `storage.*` modules should be thread-safe.

Reply:
Probably yes, but the problem is that we get access to state only in this place, and this factory gets injected into `NewState` so it should or this way or state has to be passed to factory after it's created.

Reply:
It feels a little awkward, I admit. Nevertheless, I think this is the best starting point for me to work towards [issue #5103](https://github.com/dapperlabs/flow-go/issues/5103)